### PR TITLE
enhance(publish): support unpublished links

### PR DESCRIPTION
**Pull request by kevinslin**
# Pull Request Checklist

You can go to [dendron pull requests](https://wiki.dendron.so/notes/adc39825-77a6-46cf-9c49-2642fcb4248e.html) to see full details for items in this checklist.

### Quality Assurance
- [x] add a [test](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#writing-tests) for the new feature
- [x] make sure all the existing [tests](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#running-all-tests) pass
- [x] do a spot check by running your feature with our [test workspace](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#test-workspace)
- [ ] after you submit your pull request, check the output of our [integration test](https://github.com/dendronhq/dendron/actions) and make sure all tests pass
  - NOTE: if you running mac/linux, check the windows output and vice versa if you are developing on windows

#### Special Cases
- [x] if your tests changes an existing snaphot, make sure that snapshots are [updated](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#updating-test-snapshots)
- [x] if you are adding a new language feature (graphically visible in vscode/preview/publishing), make sure that it is included in [test-workspace](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#test-workspace). We use this to manually inspect new changes and for auto regression testiing 

### Docs
- [x] Make sure that the PR title follows our [commit style](https://wiki.dendron.so/notes/adc39825-77a6-46cf-9c49-2642fcb4248e.html#commit-style)
- [x] Please summarize the feature or impact in 1-2 lines in the PR description
- [x] If your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR

### CODE REVIEW

**Comment by kevinslin** [comment]
enhance(publish): support unpublished links

Unpublished links will display the "This page has not sprouted popup" instead of a 404

**Comment by SeriousBug** [comment]
I was reviewing the code, and I did a little digging into why the ubuntu-cli tests were failing.

The issue is in these two parts:
- https://github.com/dendronhq/dendron/blob/57c275cd19b5def6055685fe70ba8b75f9774dfe/packages/engine-server/src/topics/connector.ts#L169
- https://github.com/dendronhq/dendron/blob/57c275cd19b5def6055685fe70ba8b75f9774dfe/packages/engine-server/src/topics/connector.ts#L152

If `portCreated > wsActivation` is false, then connecting to the engine enters an infinite loop because the client keeps retrying the connection but fails because of the timestamps.

I'm not sure why these timestamps don't match on Linux, but hopefully this helps with a solution.

File: packages/engine-server/src/markdown/remark/dendronPub.ts [hunk 1]
```diff
@@ -152,13 +152,16 @@ function plugin(this: Unified.Processor, opts?: PluginOpts): Transformer {
         }
 
         const copts = opts?.wikiLinkOpts;
-        if (note && opts?.transformNoPublish) {
+        if (!note && opts?.transformNoPublish) {
+          value = "403";
```
Comment (jonathanyeung) [inline packages/engine-server/src/markdown/remark/dendronPub.ts:156]: minor nit: use http status codes constants for this.

File: packages/engine-server/src/markdown/remark/dendronPub.ts [hunk 2]
```diff
@@ -152,13 +152,16 @@ function plugin(this: Unified.Processor, opts?: PluginOpts): Transformer {
...
+          addError(proc, new DendronError({ message: "no note" }));
```
Comment (jonathanyeung) [inline packages/engine-server/src/markdown/remark/dendronPub.ts:157]: can you add 403 code to the dendron error + severity?

**Review by jonathanyeung — APPROVED** [review]

**Comment by kevinslin** [comment]
> I was reviewing the code, and I did a little digging into why the ubuntu-cli tests were failing.
> 
> The issue is in these two parts:
> 
>     * https://github.com/dendronhq/dendron/blob/57c275cd19b5def6055685fe70ba8b75f9774dfe/packages/engine-server/src/topics/connector.ts#L169
> 
>     * https://github.com/dendronhq/dendron/blob/57c275cd19b5def6055685fe70ba8b75f9774dfe/packages/engine-server/src/topics/connector.ts#L152
> 
> 
> If `portCreated > wsActivation` is false, then connecting to the engine enters an infinite loop because the client keeps retrying the connection but fails because of the timestamps.
> 
> I'm not sure why these timestamps don't match on Linux, but hopefully this helps with a solution.
> 
> If I'm understanding this code correctly, instead of checking timestamps we could add a `status` endpoint to the engine and ping that to check that the port is good.

thanks for the investigation. fixed!

File: packages/engine-server/src/markdown/remark/dendronPub.ts (hunk 1 above)
Comment (kevinslin) [reply packages/engine-server/src/markdown/remark/dendronPub.ts:156]: created a separate PR for this here -> https://github.com/dendronhq/dendron/pull/1344

File: packages/engine-server/src/markdown/remark/dendronPub.ts (hunk 2 above)
Comment (kevinslin) [reply packages/engine-server/src/markdown/remark/dendronPub.ts:157]: see https://github.com/dendronhq/dendron/pull/1344

**Comment by jonathanyeung** [comment]
Shall we also close this one?

**Pull request closed without merging**